### PR TITLE
PST-003-implement-redux-+-saga

### PR DESCRIPTION
**Pull request by skrzypczyk-pawel**

### CODE REVIEW

File: src/components/AddNewTodoForm.tsx [hunk 1]
```diff
@@ -30,8 +32,10 @@ export const AddNewTodoForm: FC<Props> = () => {
   const selectedMenu = document.getElementById(
     "formSelectMenu"
   ) as HTMLSelectElement;
+  const dispatch = useDispatch();
   const [category, setCategory] = useState<Category>("none");
   const [priority, setPriority] = useState<Priority>("low");
+  const [newID, setNewID] = useState<number>(31);
```
Comment (Stared55) [inline src/components/AddNewTodoForm.tsx]: No.

File: src/components/AddNewTodoForm.tsx [hunk 2]
```diff
@@ -48,7 +52,7 @@ export const AddNewTodoForm: FC<Props> = () => {
   };
 
   const initialValues: ITodo = {
-    id: Math.floor(Math.random() * 1000) + 1,
+    id: newID,
```
Comment (Stared55) [inline src/components/AddNewTodoForm.tsx]: implement https://www.npmjs.com/package/uuid

File: src/components/AddNewTodoForm.tsx [hunk 3]
```diff
@@ -60,14 +64,19 @@ export const AddNewTodoForm: FC<Props> = () => {
 
   const additionalResetForm = () => {
     setCategory("none");
+    let ID = newID;
+    ID++;
+    setNewID(ID);
```
Comment (Stared55) [inline src/components/AddNewTodoForm.tsx]: remove this

File: src/components/AddNewTodoForm.tsx [hunk 4]
```diff
@@ -60,14 +64,19 @@ export const AddNewTodoForm: FC<Props> = () => {
...
     selectedMenu.value = "low";
   };
 
   const handleOnSubmit = (values: ITodo, actions: FormikHelpers<ITodo>) => {
     values.category = category;
     values.priority = priority;
+    values.id = newID;
```
Comment (Stared55) [inline src/components/AddNewTodoForm.tsx]: values.id = uuidv4();

File: src/store/todo/actionTypes.tsx [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export const FETCH_TODO_REQUEST = "FETCH_TODO_REQUEST";
+export const FETCH_TODO_SUCCESS = "FETCH_TODO_SUCCESS";
```
Comment (Stared55) [inline src/store/todo/actionTypes.tsx]: FETCH_TODO_RESOLVED

File: src/store/todo/actionTypes.tsx [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+export const FETCH_TODO_FAILURE = "FETCH_TODO_FAILURE";
```
Comment (Stared55) [inline src/store/todo/actionTypes.tsx]: FETCH_TODO_REJECTED

File: src/store/todo/reducer.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  todos: [],
+  error: false,
+};
+
+export default (state = initialState, action: TodoActions) => {
+  switch (action.type) {
+    case FETCH_TODO_REQUEST:
+      return {
+        ...state,
+        loading: true,
```
Comment (Stared55) [inline src/store/todo/reducer.tsx:20]: ...state,
loading: true,
error: false,

File: src/store/todo/reducer.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        ...state,
+        loading: false,
+        todos: action.payload.todos,
+        error: false,
+      };
+    case FETCH_TODO_FAILURE:
+      return {
+        ...state,
+        loading: false,
+        todos: [],
```
Comment (Stared55) [inline src/store/todo/reducer.tsx]: remove this `todos: []`

File: src/store/todo/sagas.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import { all, call, put, takeLatest } from "redux-saga/effects";
+import { fillDummyTodo } from "src/helpers";
+
+import { ITodo, RawTodo } from "src/types";
+import { fetchTodoFailure, fetchTodoSuccess } from "./actions";
+import { FETCH_TODO_REQUEST } from "./actionTypes";
+
+const todosDataURL = "https://jsonplaceholder.typicode.com/todos";
```
Comment (Stared55) [inline src/store/todo/sagas.tsx]: move into src/api/todos.ts

File: src/store/todo/sagas.tsx [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+const getTodos = () => fetch(todosDataURL);
+
+function* fetchTodoSaga() {
+  try {
+    const response: Response = yield call(getTodos);
+    const todosTasks: RawTodo[] = yield response.json();
+    const updatedTodos: ITodo[] = [];
+    todosTasks.forEach((item) => {
+      if (+item.id <= 30) {
```
Comment (Stared55) [inline src/store/todo/sagas.tsx]: better approach is slice this array instead od checking ids

File: src/store/todo/reducer.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        error: false,
```
Comment (Stared55) [inline src/store/todo/reducer.tsx]: should be string from payload

File: src/store/todo/reducer.tsx [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  FETCH_TODO_REQUEST,
+  FETCH_TODO_SUCCESS,
+  FETCH_TODO_FAILURE,
+  ADD_TODO,
+} from "./actionTypes";
+
+const initialState: TodoState = {
+  loading: false,
+  todos: [],
+  error: false,
```
Comment (Stared55) [inline src/store/todo/reducer.tsx]: error: ''

File: src/store/user/actionTypes.tsx [hunk 13]
```diff
@@ -0,0 +1,4 @@
+export const FETCH_USER_REQUEST = "FETCH_USER_REQUEST";
+export const FETCH_USER_SUCCESS = "FETCH_USER_SUCCESS";
+export const FETCH_USER_FAILURE = "FETCH_USER_FAILURE";
```
Comment (Stared55) [inline src/store/user/actionTypes.tsx]: Same pattern as above

File: src/store/todo/actions.tsx [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  FETCH_TODO_SUCCESS,
+  ADD_TODO,
+} from "./actionTypes";
+
+export const fetchTodoRequest = (): FetchTodoRequest => ({
+  type: FETCH_TODO_REQUEST,
+});
+
+export const fetchTodoSuccess = (
+  payload: FetchTodoSuccessPayload
```
Comment (Stared55) [inline src/store/todo/actions.tsx]: payload: ITodos[]

File: src/store/todo/actions.tsx [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+): FetchTodoSuccess => ({
+  type: FETCH_TODO_SUCCESS,
+  payload,
+});
+
+export const fetchTodoFailure = (
+  payload: FetchTodoFailurePayload
```
Comment (Stared55) [inline src/store/todo/actions.tsx]: payload: string

File: src/store/user/reducer.tsx [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+import {
+  FETCH_USER_REQUEST,
+  FETCH_USER_SUCCESS,
+  FETCH_USER_FAILURE,
+  ADD_USER,
+} from "./actionTypes";
+
+const initialState: UserState = {
+  users: [],
+  error: false,
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: same

File: src/store/user/reducer.tsx [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+export default (state = initialState, action: UserActions) => {
+  switch (action.type) {
+    case FETCH_USER_REQUEST:
+      return {
+        ...state,
+        pending: true,
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: 🚬

File: src/store/user/reducer.tsx [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+      };
+    case FETCH_USER_SUCCESS:
+      return {
+        ...state,
+        pending: false,
+        todos: action.payload.todos,
+        error: null,
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: error should be cleared after each pending action

File: src/store/user/reducer.tsx [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+      };
+    case FETCH_USER_FAILURE:
+      return {
+        ...state,
+        pending: false,
+        todos: [],
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: out

File: src/store/user/reducer.tsx [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        error: action.payload.error,
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: error: action.payload is enough and payload should be a string

File: src/store/user/reducer.tsx [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+      };
+    case ADD_USER:
+      const newState = state;
+      newState.users.push(action.payload);
+      return {
+        ...state,
+        todos: newState.users,
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: nice copy paste

File: src/store/user/reducer.tsx [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+      };
+    case FETCH_USER_FAILURE:
+      return {
+        ...state,
+        pending: false,
+        todos: [],
+        error: action.payload.error,
+      };
+    case ADD_USER:
+      const newState = state;
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: NO NO NO. NEVER. https://www.freecodecamp.org/news/copying-stuff-in-javascript-how-to-differentiate-between-deep-and-shallow-copies-b6d8c1ef09cd/

File: src/store/todo/reducer.tsx [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+      };
+    case FETCH_TODO_FAILURE:
+      return {
+        ...state,
+        loading: false,
+        todos: [],
+        error: false,
+      };
+    case ADD_TODO:
+      const newState = state;
```
Comment (Stared55) [inline src/store/todo/reducer.tsx]: const stateCopy = [...state];

File: src/types/redux-saga-todo.ts [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+} from "src/store/todo/actionTypes";
+import { ITodo } from "./todos";
+
+export interface TodoState {
+  loading: boolean;
+  todos: ITodo[];
+  error: boolean;
+}
+
+export interface FetchTodoSuccessPayload {
```
Comment (Stared55) [inline src/types/redux-saga-todo.ts]: Yeah it is a ok pattern but I think it is unnecesary . For those simple responses u should remove those xyPayload interfaces

File: src/types/redux-saga-user.ts [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+  FETCH_USER_FAILURE,
+  ADD_USER,
+} from "src/store/user/actionTypes";
+
+export interface UserState {
+  users: IUser[];
+  error: boolean;
+}
+
+export interface FetchUserSuccessPayload {
```
Comment (Stared55) [inline src/types/redux-saga-user.ts]: same

File: src/store/todo/reducer.tsx (hunk 8 above)
Comment (skrzypczyk-pawel) [reply src/store/todo/reducer.tsx]: Without this key, terminal shows error. We can't do map in TodoList without it.
But I understand that can cause problem with empty array. So I changed it.
```suggestion
        todos: initialState.todos,
```
It works with this change. Will it be good for you, or you prefer me to look for a different solution?

Comment (Stared55) [reply src/store/todo/reducer.tsx]: No, thats not correct. This should be handled by 31th line as it is.

File: src/store/todo/actions.tsx (hunk 14 above)
Comment (skrzypczyk-pawel) [reply src/store/todo/actions.tsx]: Do you want to change the name of this type or to change on ITodo[]?

File: src/store/todo/reducer.tsx [hunk 26]
```diff
@@ -18,27 +18,27 @@ export default (state = initialState, action: TodoActions) => {
...
         ...state,
         loading: false,
-        todos: [],
-        error: false,
+        error: action.payload,
       };
     case ADD_TODO:
-      const newState = state;
-      newState.todos.push(action.payload);
+      const stateCopy = [...state.todos];
```
Comment (Stared55) [inline src/store/todo/reducer.tsx]: `const updatedTodos = [...state.todos, action.payload];` should be enough

File: src/store/user/reducer.tsx [hunk 27]
```diff
@@ -1,43 +1,42 @@
...
       return {
         ...state,
-        pending: false,
-        todos: [],
-        error: action.payload.error,
+        loading: false,
+        error: action.payload,
       };
     case ADD_USER:
       const newState = state;
```
Comment (Stared55) [inline src/store/user/reducer.tsx]: 🚬

**Review by Stared55 — APPROVED** [review]

**Pull request merged**